### PR TITLE
Add multi-theme support to Polaris tokens

### LINKED ISSUE #9988: Establish a pattern for global theming Polaris tokens
AC:
- Establish a pattern for global theming Polaris tokens in Polaris and Web
- Consider future rollouts of high-contrast themes and merchant customization options
- Explore options that continue to avoid runtime CSS in JS theming
- Aim for a minimal upgrade path to promote a smooth transition

[Prototype](https://github.com/Shopify/polaris/pull/9877)

### PR DESCRIPTION
**Pull request by aaronccasanova**
This PR introduces multi-theme support to `@shopify/polaris-tokens` by implementing a base/variant architecture. The base theme serves as the foundation for any number of variant themes. Variant themes have two primary constraints in that they can only extend the base theme and are the only themes exposed to consumers.

One advantage of this pattern is that it simplifies system changes by flattening the relationship between themes. There is no need to consider deeply inherited themes, making updates more predictable.

<details>
  <summary>Deeply inherited vs base/variant themes</summary>

**❌ With deeply inherited themes**
```mermaid
flowchart TD
    A[Base Theme]
    A --> B[Light Theme]
    B --> C[Light High Contrast Theme]
    A --> D[Dark Theme]
    D --> E[Dark High Contrast Theme]
    A --> F[Dim Theme]
    A --> G[etc...]
```

**✅ With base/variant themes**
```mermaid
flowchart TD
    A[Base Theme]
    A --> B[Light Theme]
    A --> C[Light High Contrast Theme]
    A --> D[Dark Theme]
    A --> E[Dark High Contrast Theme]
    A --> F[Dim Theme]
    A --> G[etc...]
```
 
</details>

## Implementation Details

This PR focuses on implementing backward compatible multi-theme support for Polaris tokens. It enables the addition of new variant themes while promoting incremental adoption and ensuring compatibility with existing implementations.

### Directory Structure

The `src/themes` directory consists of the following components:

- Internal base theme and public variant themes.
- Default variant theme, which is essentially an alias of the base theme, aliased to the light theme.
- Light uplift theme, which is the first variant theme that simply ported `valueExperimental` tokens from our token groups. `valueExperimental` properties were initially introduced as a short-term solution during the SE23 upgrades, allowing token overrides by replacing `value` with `valueExperimental` at build time.
  - Note: `valueExperimental` references in token groups were kept in this PR to prevent breaking changes and maintain backward compatibility. Although these properties should be allowed to remove, a couple usages were found across the organization.
- TypeScript definitions describing the shape of themes, partial themes, theme names, etc.
- Utilities for validating themes and consistently creating theme classes/selectors.

### Stylesheet Generation

The `toStyleSheet` build step has been updated to generate a backward-compatible [styles.scss](https://unpkg.com/browse/@shopify/polaris-tokens@7.5.2/dist/css/styles.css) file. This file assigns the default variant theme custom properties (`light`) to the `:root` selector and the `light-uplift` variant theme custom properties to the `html.Polaris-Summer-Editions-2023` selector.

It should be noted that while the stylesheet remains backward compatible in this PR, the plan is to propose removing the `:root` selector in the next major release. This change aims to flatten the specificity of all variant themes since the `:where()` selector cannot currently be leveraged. For example:

- Current specificity breakdown: `:root` has a specificity of 10, `html.Polaris-Summer-Editions-2023` has a specificity of 11.
- Next specificity breakdown: `html.p-theme-light` has a specificity of 11, `html.Polaris-Summer-Editions-2023` has a specificity of 11, etc.

### Variant Theme Functionality

Variant themes override a subset of the base theme tokens by assigning them to a `partials` object. This `partials` object is then deep merged with the base theme for JavaScript access. Additionally, the `partials` object is used in the `toStyleSheet` build step to generate custom properties and insert them into the variant theme selector. Consequently, variant theme selectors only contain the subset of overridden custom properties, ensuring a minimal stylesheet.

It should be noted that if the strategy shifts from global themes to allowing nested themes, the approach would likely change to inserting all custom properties in each class. However, even in such cases, the performance impact of having a large collection of custom properties is expected to be minimal.

**Example `styles.scss` updates**

```scss
:root {
  /* Default variant theme custom properties (light) */
}

html.Polaris-Summer-Editions-2023 {
  /* Variant theme custom properties */
  /*
     Note: The above selector is special
     cased for backward compatibility. See
     below for updated theme selector structure.
  */
}

html.p-theme-light-high-contrast {
  /* Variant theme custom properties */
  /*
     Note: Variant theme selectors
     contain a subset of custom properties
     overriding the base theme
  */
}

html.p-theme-dark {/* ... */}
html.p-theme-dark-high-contrast {/* ... */}
html.p-theme-dim {/* ... */}
```

### CODE REVIEW

File: polaris-tokens/src/themes/light-uplift.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    'color-text-interactive-inverse': {value: colors.blue[8]},
+    'color-text-magic-strong': {value: colors.purple[15]},
+    'color-text-magic': {value: colors.purple[14]},
+  },
+});
+
+export const themeLightUplift = deepmerge(
+  themeBase,
+  themeLightUpliftPartial,
+) as ThemeBase;
```
Comment (lgriffee) [inline polaris-tokens/src/themes/light-uplift.ts]: Found it interesting that the variant theme gets cast here to the base theme. Not sure if we could make this clearer.

File: polaris-tokens/src/themes/types.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+export interface TokenGroupShape {
+  [tokenName: string]: TokenProperties;
+}
+
+export interface ThemeShape {
+  [tokenGroupName: string]: TokenGroupShape;
+}
+
+type ExcludeMotionKeyframes<T> = T extends `motion-keyframes-${string}`
```
Comment (sam-b-rose) [inline polaris-tokens/src/themes/types.ts:17]: What is the reason for needing to exclude `motion-keyframes-`?

File: polaris-tokens/src/themes/types.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  ? never
+  : T;
+
+export type ThemeVariantPartialShape = {
+  [TokenGroupName in keyof Omit<ThemeBase, 'breakpoints'>]?: {
```
Comment (sam-b-rose) [inline polaris-tokens/src/themes/types.ts:22]: Would we ever want to have a theme with breakpoint overrides? Also might be worth keeping in here

**Review by lgriffee — APPROVED** [review]
LGTM! 💯🚢✨🎉 This is awesome Aaron! 🙌🏻

**Review by sam-b-rose — COMMENTED** [review]
This looks excellent @aaronccasanova great job retaining maximum TS info when creating the partials, nice utility functions, and great use of CSS terminology for naming things 👍

File: polaris-tokens/src/themes/light-uplift.ts (hunk 1 above)
Comment (aaronccasanova) [reply polaris-tokens/src/themes/light-uplift.ts]: Good call 👍 We can configure the `deepmerge` type parameters to properly intersect `ThemeBase` and `typeof  themeLightUplift` e.g.

```ts
export const themeLightUplift = deepmerge<ThemeBase, typeof themeLightUplift>(
  themeBase,
  themeLightUpliftPartial,
);
```

However, I find it bit verbose and the collective type utilities make it basically equivalent to the cast.

Perhaps, we can remove ambiguity with a `createThemeVariant` util. Thoughts?

```ts
export const themeLightUplift = createThemeVariant(themeLightUpliftPartial);
```

Comment (lgriffee) [reply polaris-tokens/src/themes/light-uplift.ts]: Gottcha! I like your idea of creating a `createThemeVariant` util 👍🏻 Also dig the added benefit of ensuring deepmerge is configured correctly each time.

File: polaris-tokens/src/themes/types.ts (hunk 2 above)
Comment (aaronccasanova) [reply polaris-tokens/src/themes/types.ts:17]: Good question. While it's possible to create themed keyframes, we don't currently have a need for them and I decided to exclude them to reduce complexity. A brief summary and breakdown of how to achieve themed keyframes is outlined in the [Tradeoffs section of the proposal](https://github.com/Shopify/polaris/issues/9988#issuecomment-1675507759).

Comment (sam-b-rose) [reply polaris-tokens/src/themes/types.ts:17]: Makes sense, thanks for documenting and excellent reasoning 👍 in favor of keeping it as-is

File: polaris-tokens/src/themes/types.ts (hunk 3 above)
Comment (sam-b-rose) [reply polaris-tokens/src/themes/types.ts:22]: Disregard. I see why this wouldn't work since our breakpoints are tied to SCSS variables. 👍

Comment (aaronccasanova) [reply polaris-tokens/src/themes/types.ts:22]: I don't think `breakpoints` are stable enough to support theming. There are too many hard coded breakpoints across the org where any change requires automated and manual migrations (and deep product area expertise).

File: polaris-tokens/src/themes/light-uplift.ts (hunk 1 above)
Comment (aaronccasanova) [reply polaris-tokens/src/themes/light-uplift.ts]: Updated! https://github.com/Shopify/polaris/pull/10103/commits/24b86f27573a15de86100c838cbe59d7007dd6e2

File: polaris-tokens/scripts/toStyleSheet.ts [hunk 4]
```diff
@@ -1,90 +1,67 @@
...
-      ([token]) =>
-        token.startsWith('motion-keyframes') || token.startsWith('keyframes'),
-    )
+    .filter(([token]) => token.startsWith('motion-keyframes'))
     .map(([token, {value}]) => `@keyframes p-${token}${value}`)
     .join('');
 }
 
-export async function toStyleSheet(metadata: Metadata) {
+export async function toStyleSheet(
```
Comment (aaronccasanova) [inline polaris-tokens/scripts/toStyleSheet.ts:44]: Validated there is no diff between the new and old stylesheet:
Screenshot 2023-08-18 at 3 25 33 PM

**Review by kyledurand — APPROVED** [review]
👏👏👏

**Review by sam-b-rose — APPROVED** [review]

**Pull request merged**